### PR TITLE
[IMP] sale: move amount_to_invoice computation to SO line

### PR DESCRIPTION
**Pull request by epictete**
In 16.2, the amount to invoice was introduced to effectively compute the total receivable of a partner.
The goal was to take into account the confirmed SOs that are not invoiced yet, in order to trigger the partner's credit limit warning.
Since then:
- The total receivable of a partner = unpaid posted invoices + SOs amount to invoice.
- Where the amount to invoice = SO total - confirmed down payments.

The issue with that method is that we go against the Sales philosophy, which considers that a SOL (Sales Order Line) is invoiced as soon as the invoiced quantity (whatever the invoice status) equals the ordered quantity.
That can lead to incoherent situations.
For example:
Let's imagine a SO with 2 products:
- 3 desks at €100
- 5 chairs at €50.

We invoice 3 desks at 90€ because of a price change and 5 chairs at 50€.
Sales will consider the SO to be done ("nothing to invoice"), whereas Accounting consider that there's still €30 to invoice.
The bottom line is that:
- We should consider the amount to invoice per SOL and not per SO total.
- We should remain consistent with the Sales methodology, to avoid weird cases as demonstrated before.

task-3764582

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by Feyensv — COMMENTED** [review]
3 new stored fields on sale order lines ?
Isn't there a simpler way to manage all that ?
I doubt we want to add all that only for a side feature like the credit limit :thinking:

**Comment by william-andre** [comment]
@Feyensv This was my first reaction too. But considering that there are already fields with the same dependencies that are stored and need to be written, we won't lose anything when writing.
On the other hand, if we don't store it, we will lose time when reading.

**Comment by Feyensv** [comment]
> @Feyensv This was my first reaction too. But considering that there are already fields with the same dependencies that are stored and need to be written, we won't lose anything when writing. On the other hand, if we don't store it, we will lose time when reading.

when reading ? All I see is that you need the order fields to be right, but I don't see any need for the SOL fields, they could be simple getters called by the SO computes, and their dependencies could be moved to the SO computed field dependencies.

**Comment by william-andre** [comment]
> but I don't see any need for the SOL fields

My argument still stands that it doesn't cost anything to store it since we compute other fields at the same time.

**Comment by Feyensv** [comment]
* Nightmare at upgrade (I don't see any script, so I guess we simply have to compute it on ALL the SO/SOL in existing databases :fire:)
* If the point is 'Oh, there is already some load, let's add more', then maybe that we should consider the opposite and stop storing the `amount_to_invoice` ?  Honestly, computing so many fields at all times for an optional feature like that doesn't seem worth the gain (might be worth moving it in a separate module maybe?).

File: addons/sale/models/sale_order_line.py [hunk 1]
```diff
@@ -889,7 +920,7 @@ def _can_be_invoiced_alone(self):
         return self.product_id.id != self.company_id.sale_discount_product_id.id
 
     @api.depends('invoice_lines', 'invoice_lines.price_total', 'invoice_lines.move_id.state', 'invoice_lines.move_id.move_type')
-    def _compute_untaxed_amount_invoiced(self):
+    def _compute_invoiced_amounts(self):
```
Comment (Feyensv) [inline addons/sale/models/sale_order_line.py:923]: `_compute_untaxed_amount_invoiced` is overridden in `pos_sale` :eyes:

File: addons/sale/tests/test_sale_to_invoice.py [hunk 2]
```diff
@@ -1101,3 +1101,52 @@ def test_amount_to_invoice_one_line_multiple_so(self):
...
+
+        so.action_confirm()
+        sol_prod_deliver.write({'qty_delivered': 5.0})
+
+        ctx = {
+            'active_model': 'sale.order',
+            'active_ids': [so.id],
+            'active_id': so.id,
+            'default_journal_id': self.company_data['default_journal_sale'].id,
+        }
```
Comment (william-andre) [inline addons/sale/tests/test_sale_to_invoice.py:1132]: I don't think that the context is relevant?
It would be more explicit to give `sale_order_ids` to the `create` values.

File: addons/sale/models/sale_order.py [hunk 3]
```diff
@@ -216,8 +216,8 @@ def _rec_names_search(self):
     amount_untaxed = fields.Monetary(string="Untaxed Amount", store=True, compute='_compute_amounts', tracking=5)
     amount_tax = fields.Monetary(string="Taxes", store=True, compute='_compute_amounts')
     amount_total = fields.Monetary(string="Total", store=True, compute='_compute_amounts', tracking=4)
-    amount_to_invoice = fields.Monetary(string="Amount to invoice", store=True, compute='_compute_amount_to_invoice')
-    amount_invoiced = fields.Monetary(string="Already invoiced", compute='_compute_amount_invoiced')
+    amount_to_invoice = fields.Monetary(string="Un-invoiced Balance", compute='_compute_invoicing_amounts', store=True)
+    amount_invoiced = fields.Monetary(string="Already invoiced", compute='_compute_invoicing_amounts', store=True)
```
Comment (william-andre) [inline addons/sale/models/sale_order.py:220]: This needs an upgrade script. Do we actually need to store?

File: addons/sale/models/sale_order_line.py [hunk 4]
```diff
@@ -266,12 +270,19 @@ class SaleOrderLine(models.Model):
 
     untaxed_amount_invoiced = fields.Monetary(
         string="Untaxed Invoiced Amount",
-        compute='_compute_untaxed_amount_invoiced',
+        compute='_compute_invoiced_amounts',
+        store=True)
+    amount_invoiced = fields.Monetary(
```
Comment (william-andre) [inline addons/sale/models/sale_order_line.py:275]: Do we need to store this?

File: addons/sale/models/sale_order.py (hunk 3 above)
Comment (epictete) [reply addons/sale/models/sale_order.py:220]: No need to store them indeed :slightly_smiling_face:

File: addons/sale/models/sale_order_line.py (hunk 4 above)
Comment (epictete) [reply addons/sale/models/sale_order_line.py:275]: Yes, because both `untaxed_amount_invoiced` and `amount_invoiced` share the same compute method, and `untaxed_amount_invoiced` was already stored (used in an SQL query in `sale_report.py`)

Comment (william-andre) [reply addons/sale/models/sale_order_line.py:275]: > Yes, because both `untaxed_amount_invoiced` and `amount_invoiced` share the same compute method

I'll change my question then :smile: 
Does it have to be the same compute function?

Comment (epictete) [reply addons/sale/models/sale_order_line.py:275]: Both fields share the same dependencies and computation logic, so it made sense to use the same compute method, no?

Comment (william-andre) [reply addons/sale/models/sale_order_line.py:275]: Well... yes. But also no.
* because we need one stored but not the other
* generally, each stored field needs needs it's own compute method

______________
Stored compute fields need to have their own computed method, otherwise
```python
class MyModel(models.Model):
    _name = 'my.model'
    fieldA = fields.Char(compute='_compute_fields', store=True, readonly=False)
    fieldB = fields.Char(compute='_compute_fields', store=True, readonly=False)

    def _compute_fields(self):
        for record in self:
            record.fieldA = 'a'
            record.fieldB = 'b'
```
```python
env['my.model'].create({'fieldA': 'c'})
# `fieldA` will be 'a' instead of 'c' because 
# `_compute_fields` needs to be called to compute `fieldB`
```

**Review by william-andre — APPROVED** [review]
@robodoo r+

File: addons/sale/models/sale_order_line.py [hunk 5]
```diff
@@ -1021,6 +1060,19 @@ def _compute_untaxed_amount_to_invoice(self):
 
             line.untaxed_amount_to_invoice = amount_to_invoice
 
+    @api.depends('discount', 'price_total', 'product_uom_qty', 'qty_delivered', 'qty_invoiced_posted')
+    def _compute_amount_to_invoice(self):
+        to_compute = self.company_id.account_use_credit_limit
+        for line in self:
```
Comment (william-andre) [inline addons/sale/models/sale_order_line.py:1066]: `self.company_id` can be multiple companies
```suggestion
        to_compute = self.filtered('company_id.account_use_credit_limit')
        (self - to_compute).amount_to_invoice = 0
        for line in to_compute:
```

File: addons/sale/models/sale_order_line.py [hunk 6]
```diff
@@ -1021,6 +1060,19 @@ def _compute_untaxed_amount_to_invoice(self):
 
             line.untaxed_amount_to_invoice = amount_to_invoice
 
+    @api.depends('discount', 'price_total', 'product_uom_qty', 'qty_delivered', 'qty_invoiced_posted')
+    def _compute_amount_to_invoice(self):
+        to_compute = self.company_id.account_use_credit_limit
```
Comment (william-andre) [inline addons/sale/models/sale_order_line.py:1065]: Since it's not in the view, I wouldn't filter out `account_use_credit_limit`.

In practice, this field will only be computed when used in other functions, and thos functions already filter based on the credit limit

**Review by william-andre — COMMENTED** [review]
I feel like this is messy for little to no reason.

The code could be more readable if we saved the amounts and quantities saved in a JSON field, with the keys being the state of the invoice.
From there, everything should be easy to compute...

File: addons/sale/tests/test_credit_limit.py [hunk 7]
```diff
@@ -36,7 +36,7 @@ def test_credit_limit_multi_company(self):
         company2 = self.company_data_2['company']
 
         # Activate the Credit Limit feature and set a value for partner_a.
-        self.env.company.account_use_credit_limit = True
+        company2.account_use_credit_limit = True
```
[the remark this answers is not in the document]
Comment (epictete) [reply addons/sale/tests/test_credit_limit.py:39]: I slightly updated the test (and comments) to make it clearer.

File: addons/l10n_it_edi_doi/models/sale_order_line.py [hunk 8]
```diff
@@ -1,31 +0,0 @@
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import api, fields, models
-
-
-class SaleOrderLine(models.Model):
-    _inherit = "sale.order.line"
-
-    qty_invoiced_posted = fields.Float(
```
[the remark this answers is not in the document]
Comment (epictete) [reply addons/l10n_it_edi_doi/models/sale_order_line.py:9]: Yes, the field is now moved to `sale`, so `_l10n_it_edi_doi_get_amount_not_yet_invoiced` is still going to work.

File: addons/sale/models/sale_order_line.py (hunk 5 above)
Comment (epictete) [reply addons/sale/models/sale_order_line.py]: I think it makes more sense to look at the selected company, instead of the company of the SO.
Since all these fields are meant to be used in the context of the credit limit, it makes sense to only do the computations when the setting is actually activated on the selected company, no?
Plus `credit_to_invoice` can only rely on the selected company since there is no company associated with partners, so it seemed logical to use the same condition everywhere :blush:

File: addons/sale/models/sale_order_line.py (hunk 6 above)
Comment (epictete) [reply addons/sale/models/sale_order_line.py]: It was just an extra precaution, to be safe, but if you think it's not necessary, I can remove it :wink:

File: addons/sale/models/sale_order_line.py [hunk 9]
```diff
@@ -1021,6 +1060,20 @@ def _compute_untaxed_amount_to_invoice(self):
 
             line.untaxed_amount_to_invoice = amount_to_invoice
 
+    @api.depends('discount', 'price_total', 'product_uom_qty', 'qty_delivered', 'qty_invoiced_posted')
+    def _compute_amount_to_invoice(self):
+        if not self.company_id.account_use_credit_limit:
+            self.amount_to_invoice = 0.0
```
Comment (william-andre) [inline addons/sale/models/sale_order_line.py:1066]:
```suggestion
```

**Review by william-andre — APPROVED** [review]
@robodoo r+

**Comment by william-andre** [comment]
@robodoo rebase-ff

**Comment by william-andre** [comment]
@robodoo r+
